### PR TITLE
[TRTLLM-14287][feat] Qwen Image CFG parallelism support

### PR DESCRIPTION
**Pull request by yibinl-nvidia**
## Summary by CodeRabbit

* **New Features**
  * Added distributed CFG parallel generation for Qwen-Image pipelines.
  * Improved true classifier-free guidance with norm-based noise rescaling.
  * Added support for flexible negative prompt inputs, including validation for batch sizes.

* **Bug Fixes**
  * Improved guidance behavior when negative prompts are omitted.

* **Tests**
  * Added coverage for prompt handling, distributed CFG state, input selection, and guidance calculations.

## Description
| CFG=1 | CFG=2 |
|---|---|
| qwen_cfg1_negative_prompt_direct_output | qwen_cfg2_negative_prompt_direct_output |

Measured with explicit negative_prompt="", 1328x1328, 50 steps. 

  | Case | GPUs | Generation time |                                                                                                           
  |---|---:|---:|                                                                                                                             
  | CFG=1 true CFG, positive + negative serial | 1 | 16.22s |
  | CFG=2 CFG parallel | 2 | 8.51s |
                                   
  CFG=2 is about 1.91x faster. LPIPS is 0.00000000.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 1]
```diff
@@ -369,6 +370,64 @@ def _decode_latents(self, latents: torch.Tensor, height: int, width: int) -> tor
...
+            if cfg_size != 2:
+                raise ValueError(
+                    f"Qwen-Image CFG parallel only supports cfg_size=2 "
+                    f"(cond/uncond), got cfg_size={cfg_size}"
+                )
+            if not dist.is_initialized():
+                raise RuntimeError("Qwen-Image CFG parallel requires torch.distributed")
+            if cfg_rank == 0:
+                logger.info("CFG parallel: cfg_size=2")
+        return do_cfg_parallel, cfg_size, cfg_rank, cfg_pg
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: modified.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 2]
```diff
@@ -154,13 +157,16 @@ def _run_warmup(self, height: int, width: int, num_frames: int, steps: int) -> N
         ``forward`` path so CUDA graphs / torch.compile / VAE kernels
         all get triggered with the runtime shape.
         """
+        vgm = self.pipeline_config.visual_gen_mapping
+        cfg_size = vgm.cfg_size if vgm else 1
+        warmup_true_cfg_scale = 4.0 if cfg_size > 1 else 1.0
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: What does the "true/real" of the "true/real CFG" implies in Qwen-Image implementation?

Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: true CFG means “actual cond/uncond CFG with a negative prompt path.” That is different from models like FLUX, where guidance_scale can mean an embedded/distilled guidance value passed into the transformer, not a second negative-prompt forward pass.  I get that this is very confusing, updated PR title to make it a bit clear.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 3]
```diff
@@ -60,7 +62,12 @@ def _calculate_shift(
 }
 
 
-@register_pipeline("QwenImagePipeline")
+@register_pipeline(
+    "QwenImagePipeline",
+    hf_ids=[
+        "Qwen/Qwen-Image",
+    ],
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: nit: maybe a help/doc for the pipeline.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 2 above)
Comment (zhenhuaw-me) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: can we rename the "true cfg" vars to avoid confusion?

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 4]
```diff
@@ -154,13 +162,16 @@ def _run_warmup(self, height: int, width: int, num_frames: int, steps: int) -> N
...
+        cfg_size = vgm.cfg_size if vgm else 1
+        warmup_true_cfg_scale = 4.0 if cfg_size > 1 else 1.0
         with torch.no_grad():
             self.forward(
                 prompt="warmup",
                 height=height,
                 width=width,
                 num_inference_steps=max(steps, 2),
-                true_cfg_scale=1.0,
+                true_cfg_scale=warmup_true_cfg_scale,
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: can we rename the "true cfg" vars to avoid confusion?

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 5]
```diff
@@ -369,6 +380,78 @@ def _decode_latents(self, latents: torch.Tensor, height: int, width: int) -> tor
...
+        prompt_embeds_mask: torch.Tensor,
+        neg_prompt_embeds: torch.Tensor,
+        neg_prompt_embeds_mask: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if cfg_rank == 0:
+            return prompt_embeds, prompt_embeds_mask
+        return neg_prompt_embeds, neg_prompt_embeds_mask
+
+    @staticmethod
+    def _combine_true_cfg(
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: can we rename the "true cfg" vars to avoid confusion?

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 6]
```diff
@@ -369,6 +380,78 @@ def _decode_latents(self, latents: torch.Tensor, height: int, width: int) -> tor
...
+        noise_pred: torch.Tensor, neg_noise_pred: torch.Tensor, true_cfg_scale: float
+    ) -> torch.Tensor:
+        comb = neg_noise_pred + true_cfg_scale * (noise_pred - neg_noise_pred)
+        cond_norm = torch.norm(noise_pred, dim=-1, keepdim=True)
+        noise_norm = torch.norm(comb, dim=-1, keepdim=True)
+        return comb * (cond_norm / noise_norm)
+
+    def _cfg_parallel_state(
+        self, do_true_cfg: bool
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: can we rename the "true cfg" vars to avoid confusion?

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 5 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: Switched to `negative_prompt_cfg` instead.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 6 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: Switched to `negative_prompt_cfg` instead.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 4 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: Switched to `negative_prompt_cfg` instead.

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 3 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: added a doc change to `docs/source/models/supported-models.md`

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py [hunk 7]
```diff
@@ -369,6 +375,78 @@ def _decode_latents(self, latents: torch.Tensor, height: int, width: int) -> tor
...
+        prompt_embeds_mask: torch.Tensor,
+        neg_prompt_embeds: torch.Tensor,
+        neg_prompt_embeds_mask: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if cfg_rank == 0:
+            return prompt_embeds, prompt_embeds_mask
+        return neg_prompt_embeds, neg_prompt_embeds_mask
+
+    @staticmethod
+    def _combine_negative_prompt_cfg(
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py:412]: not sure why emphasize "negative prompt" concept (IIUC CFG always uses neg prompt)

**Review by zhenhuaw-me — APPROVED** [review]

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 7 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py:412]: Because CFG unconditional path may be implemented as an empty prompt, a negative prompt, embedded guidance, or model-specific guidance. In Qwen-Image specifically, the second CFG branch is driven by negative_prompt, so names like `negative_prompt_cfg_scale` make it clear this path only applies when we are doing the positive-vs-negative-prompt CFG flow.

**Review by chang-l — APPROVED** [review]

File: tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py (hunk 2 above)
Comment (yibinl-nvidia) [reply tensorrt_llm/_torch/visual_gen/models/qwen_image/pipeline_qwen_image.py]: done.
